src/envelope.rs

```diff
@@ -11,31 +12,82 @@ pub(crate) struct Envelope {
 
 #[derive(Debug)]
 pub(crate) enum Protocol {
+    Tcp(Segment),
     Udp(Datagram),
 }
 
 #[derive(Debug)]
 pub(crate) struct Datagram(pub Bytes);
 
+/// This is a parody of TCP.
+///
+/// We implement just enough to make it realistic, but we skip a ton of
+/// complexity (e.g. checksums, flow control, etc).
+#[derive(Debug)]
+pub(crate) enum Segment {
+    Syn(Syn),
+    Data(u64, StreamData),
+    Fin(u64),
+    Rst,
+}
+
+#[derive(Debug)]
+pub(crate) struct StreamData(pub Bytes);
+
+impl StreamData {
+    pub(crate) fn eof() -> Self {
+        Self(Bytes::new())
```

Comment:
I don't know how, but I know this will byte (sic) us.

Reply:
I adding another layer of modeling to avoid this.